test/new_relic/agent/instrumentation/rails/active_job_subscriber.rb

```diff
@@ -0,0 +1,201 @@
+
+        time = Time.now.to_f
+        SUBSCRIBER.finish(NAME, ID, {})
+        segment = txn.segments.last
+
+        assert_in_delta time, segment.end_time
+        assert_predicate(segment, :finished?)
+      end
+    end
+
+    def test_finish_with_exception_payload
+      skip_unless_minitest5_or_above
+
+      exception_object = StandardError.new
+      noticed = false
+      segment = MiniTest::Mock.new
+      segment.expect :notice_error, nil, [exception_object]
+      SUBSCRIBER.stub(:pop_segment, segment, [ID]) do
+        SUBSCRIBER.finish(NAME, ID, {exception_object: exception_object})
+      end
+
+      segment.verify
+    end
+
+    def test_finish_with_exception_raised
+      logger = MiniTest::Mock.new
+
+      NewRelic::Agent.stub :logger, logger do
+        logger.expect :error, nil, [/Error during .* callback/]
+        logger.expect :log_exception, nil, [:error, RuntimeError]
+
+        in_transaction do |txn|
+          SUBSCRIBER.state.stub :is_execution_traced?, -> { raise 'kaboom' } do
+            SUBSCRIBER.finish(NAME, ID, {})
+          end
+
+          assert_equal 1, txn.segments.size
+        end
+      end
+      logger.verify
+    end
+
+    def test_finish_when_not_tracing
+      state = MiniTest::Mock.new
+      state.expect :is_execution_traced?, false
+
+      SUBSCRIBER.stub :state, state do
+        assert_nil SUBSCRIBER.finish(NAME, ID, {})
+      end
+    end
+
+    def test_finish_segment_when_a_segment_does_not_exist
+      SUBSCRIBER.stub :pop_segment, nil, [ID] do
+        assert_nil SUBSCRIBER.send(:finish_segment, ID, {})
+      end
+    end
+
+    # perform.active_job
+    def test_an_actual_job_event_perform
+      job = TestJob.new
+      in_transaction do |txn|
+        job.perform_now
+        validate_transaction(txn, 'perform')
+      end
+    end
+
+    # enqueue_at.active_job
+    def test_an_actual_job_event_enqueue_at
+      in_transaction do |txn|
+        TestJob.set(wait_until: 7.hours.from_now).perform_later
+        validate_transaction(txn, 'enqueue_at')
+      end
+    end
+
+    # enqueue.active_job
+    def test_an_actual_job_event_enqueue
+      in_transaction do |txn|
+        TestJob.perform_later
+        validate_transaction(txn, 'enqueue')
+      end
+    end
+
+    # perform_start.active_job
+    # enqueue_retry.active_job
+    def test_an_actual_job_event_retry
+      in_transaction do |txn|
+        TestJob.perform_now(RetryMe)
+        validate_transaction(txn, %w[enqueue_retry perform_start])
+      end
+    end
+
+    # discard.active_job
+    def test_an_actual_job_event_retry
+      in_transaction do |txn|
+        TestJob.perform_now(DiscardMe)
+        validate_transaction(txn, 'discard')
+      end
+    end
+
+    # TODO: test for retry_stopped
+    # retry_stopped.active_job
+    # def test_an_actual_job_event_retry
+    #   in_transaction do |txn|
+    #     # ???
+    #     validate_transaction(txn, 'retry_stopped')
+    #   end
+    # end
```

Comment:
Do we still need this?

Reply:
It's the 1 out of 7 that we don't have a complete integration style test for. From reviewing the ActiveJob code to see how it works, I'm comfortable that our existing instrumentation will process the events just fine. But given that the test file has 6 other individual tests with event name comment labels, I figured I'd leave the seventh behind to communicate that we know about it. Happy to remove and ignore it and/or create a backlogged story for it if you'd prefer one of those other approaches.

Reply:
Updated the TODO comment so that it no longer looks like a forgotten bit of coding.

Addressed with [6e2aaf5](https://github.com/newrelic/newrelic-ruby-agent/pull/1761/commits/6e2aaf5f00b3447670f68fe508a2d7c52bd6263d)

Reply:
Thanks, James. Could you create a story and add it to the Improve Rails Instrumentation 2.0 milestone?

Reply:
Sure thing. Created #1764

Reply:
Thank you!